### PR TITLE
Host API

### PR DESCRIPTION
**Pull request by YaronWittenstein**
# Motivation

Ergonomic API to be used when implementing Apps in Rust (for example - in upcoming Vault app).

* Removed the `Host Context` (mapping between host fields indexes to blobs),
instead, the Host (i.e `go-spacemesh`) will inject these fields as imports to SVM.
For example: if SVM running app will want to know its `sender` address 
it will call an imported function named: `host_sender` (under wasm namespace `host`).

* Mock Host API implementation.

* Testing the external Host API (`ExtHost`) will be done when testing the Vault app.

### CODE REVIEW

File: crates/svm-sdk/src/host/ext.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+extern "C" {
+    /// Returns the memory offset where the transaction's input `calldata` starts.
+    fn svm_calldata_offset() -> u32;
+
+    /// Returns the transaction's input `calldata` byte-length.
+    fn svm_calldata_len() -> u32;
+
+    /// Signals to SVM that the current running transaction output (a.k.a `returndata`)
+    /// lays out in memory starting from offset `offset` and its byte-length is `length`.
+    fn svm_returndata(offset: u32, length: u32);
```
Comment (moshababo) [inline crates/svm-sdk/src/host/ext.rs:33]: Calling it `svm_set_returndata`, same as in the `Host` trait, will make more sense.

File: crates/svm-sdk/src/host/ext.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+#[inline]
+fn host() -> MutexGuard<'static, InnerHost> {
+    HOST.lock().unwrap()
+}
+
+/// Implements the `Host` trait.
+/// Its methods delegate work to the singleton `InnerHost` instance
+/// which also implements the `Host` trait and contains the actual implementation of the `Host` trait.
+///
+/// In order to get access to this singleton instance one the API user should use `ExtHost::instance()`
```
Comment (moshababo) [inline crates/svm-sdk/src/host/ext.rs:93]: "singleton instance **one** the API" - redundant word?

**Review by moshababo — APPROVED** [review]

File: crates/svm-sdk/src/host/ext.rs (hunk 1 above)
Comment (YaronWittenstein) [reply crates/svm-sdk/src/host/ext.rs:33]: np (I thought it'll be nicer to keep it short but not critical)

File: crates/svm-sdk/src/host/ext.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+/// ## Host Imports
+///
+/// WASM imports under namespace `host` for SVM programs targeting Spacemesh Full-Node (i.e `go-spacemesh`).
+/// Each running SVM programs under Spacemesh platform can assume their existence.
+///
+/// If other blockchain projects will want to take SVM and use it for their purposes then they
+/// should bring their own `host imports`.
+#[link(wasm_import_module = "host")]
```
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/svm-sdk/src/host/ext.rs:48]: good idea

File: crates/svm-sdk/src/host/ext.rs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    /// Returns the transaction's input `calldata` byte-length.
+    fn svm_calldata_len() -> u32;
+
+    /// Signals to SVM that the current running transaction output (a.k.a `returndata`)
+    /// lays out in memory starting from offset `offset` and its byte-length is `length`.
+    fn svm_returndata(offset: u32, length: u32);
+
+    /// Sends to SVM the logging message that starts
+    /// at memory offset `offset` (of byte-length `length`)
+    /// and it's associated message code (for signaling errors severity such as `trace/info/error` etc.)
```
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/svm-sdk/src/host/ext.rs:37]: I don't think we need to restrict it here.
I think that each app can come up with its own error codes.

File: crates/svm-sdk/src/host/ext.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (YaronWittenstein) [reply crates/svm-sdk/src/host/ext.rs:33]: in general, this should result in the wasm instance panicking...

**Pull request merged**